### PR TITLE
Implement cookie-based session-like behavior

### PR DESCRIPTION
**Pull request by ajacksified**
Also bump to latest horse/horse-react once https://github.com/reddit/horse/pull/1 and https://github.com/reddit/horse-react/pull/1 are merged.

:eyeglasses: @btholt, and also @spladug for interest and @JordanMilne in case I did something dumb

### CODE REVIEW

File: src/oauth.es6.js [hunk 1]
```diff
@@ -27,15 +33,15 @@ var oauthRoutes = function(app) {
     var redirectURI = OAuth2.authCode.authorizeURL({
       redirect_uri: redirect,
       scope: 'history,identity,mysubreddits,read,subscribe,vote,submit,save',
-      state: req.csrfToken(),
+      state: 'lambeosaurus',
```
Comment (ajacksified) [inline src/oauth.es6.js]: Going to reimplement CSRF in a different PR.

File: src/oauth.es6.js [hunk 2]
```diff
@@ -47,26 +53,23 @@ var oauthRoutes = function(app) {
...
-      }
+    token = OAuth2.accessToken.create(result);
 
-      var user = yield app.V1Api(req).users.get(options);
+    this.cookies.set('token', token.token.access_token, { signed: true });
 
-      this.session.user = user;
-      this.redirect(req.session.redirect || '/');
+    var options = {
+      user: 'me', //the current oauth api doesn't return userid. :(
```
Comment (btholt) [inline src/oauth.es6.js]: Can we bug platform to fix this? Seems like an oversight.

File: src/routes.jsx [hunk 3]
```diff
@@ -261,9 +262,9 @@ function routes(app) {
       text: ctx.body.text
     });
 
-    var api = app.V1Api(this);
+    var api = app.V1Api(props.token);
 
-    if (!(this.session && this.session.token && this.session.token.access_token)) {
+    if (!(this.token)) {
```
Comment (btholt) [inline src/routes.jsx]: Probably doesn't need to be wrapped in extra parens.

File: src/views/pages/listing.jsx [hunk 4]
```diff
@@ -71,7 +72,7 @@ class ListingPage extends React.Component {
 
     if (!loading) {
       listingElement = (
-        <Listing listing={listing} single={ true } session={ session } api={api} expanded={ true } />
+        <Listing listing={ listing } single={ true } user={ user } token={ token } api={ api } expanded={ true } />
```
Comment (btholt) [inline src/views/pages/listing.jsx]: On these long attributes declaration, I think it's stylistically nice to put each on a new line.

```javascript
<Listing
  listing={ listing }
  single={ true }
  user={ user }
  token={ token }
  api={ api }
  expanded={ true }
/>
```

File: src/oauth.es6.js (hunk 2 above)
Comment (ajacksified) [inline src/oauth.es6.js]: I'm hoping that we end up implementing JWTs in OAuth and get a signed json blob back. We have an api authentication meeting on Friday to discuss it.

File: src/routes.jsx (hunk 3 above)
Comment (ajacksified) [inline src/routes.jsx]: MORE PARENS)))))))

File: src/views/pages/listing.jsx (hunk 4 above)
Comment (ajacksified) [inline src/views/pages/listing.jsx]: Fair enough - I'll make a style pass on the jsx files as a separate commit after this review.

File: src/oauth.es6.js (hunk 2 above)
Comment (JordanMilne) [inline src/oauth.es6.js]: Currently you need the `identity` OAuth scope to learn the user's identity, how would this interact with that?

Comment (ajacksified) [inline src/oauth.es6.js]: Do you mean JWTs, or do you mean what I have now? I'm currently getting back all of the data I need.

With JWTs, when you OAuth response returns, you'd get (at minimum) an object with only user id, and with proper scopes, you'd get more data in your user object.

File: src/oauth.es6.js [hunk 5]
```diff
@@ -47,26 +53,23 @@ var oauthRoutes = function(app) {
...
+      headers: {
+        Authorization: 'bearer ' + token.token.access_token,
+      },
     }
+
+    var user = yield app.V1Api(token.token.access_token).users.get(options);
+
+    this.cookies.set('user', JSON.stringify(user), { signed: true });
```
Comment (JordanMilne) [inline src/oauth.es6.js]: Hmm, is it possible to have `cookies.set()` optionally add a `Secure` flag to all cookies? Prod's going to be HTTPS-only so they'll all need one.

Comment (ajacksified) [inline src/oauth.es6.js]: Sure - I'll read from the config file to determine if this should be secure.

File: src/config.es6.js [hunk 6]
```diff
@@ -22,7 +22,7 @@ var config = {
   googleAnalyticsId: process.env.GOOGLE_ANALYTICS_ID,
   googleAnalyticsDomain: process.env.GOOGLE_ANALYTICS_DOMAIN,
 
-  keys: ['lambeosaurus'],
+  keys: [ process.env.SERVER_SIGNED_COOKIE_KEY || 'lambeosaurus' ],
```
Comment (JordanMilne) [inline src/config.es6.js]: This should complain if it falls back to the placeholder key

/me remembers when we accidentally broke CSRF prevention on www. for a week

**Comment by JordanMilne** [comment]
:nail_care: Just a couple nits, looks good!

**Comment by btholt** [comment]
:nail_care: A few nits from here as well. There are a few of those really long `<Component … />` tags.

File: src/config.es6.js (hunk 6 above)
Comment (ajacksified) [inline src/config.es6.js]: Done! Throws out a `console.warn`.

**Pull request merged**